cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -477,36 +607,58 @@ public boolean apply(RunRecordMeta input) {
     Map<ProgramRunId, RunRecordMeta> programRunIdMap = new LinkedHashMap<>();
     for (Map.Entry<MDSKey, RunRecordMeta> entry : keymap.entrySet()) {
       ProgramId programId = getProgramID(entry.getKey());
-      programRunIdMap.put(new ProgramRunId(programId.getNamespace(), programId.getApplication(),
-                                           programId.getType(), programId.getProgram(), entry.getValue().getPid()),
-                          entry.getValue());
+      programRunIdMap.put(programId.run(entry.getValue().getPid()), entry.getValue());
     }
     return programRunIdMap;
   }
 
-  private  Map<MDSKey, RunRecordMeta> getHistoricalRuns(ProgramId programId, ProgramRunStatus status,
-                                                final long startTime, final long endTime, int limit,
-                                                @Nullable Predicate<RunRecordMeta> filter) {
-    MDSKey historyKey = getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programId).build();
+  private Map<ProgramRunId, RunRecordMeta> getHistoricalRuns(@Nullable ProgramId programId, ProgramRunStatus status,
+                                                             final long startTime, final long endTime, int limit,
+                                                             @Nullable Predicate<RunRecordMeta> filter) {
+    if (programId == null || !programId.getVersion().equals(ApplicationId.DEFAULT_VERSION)) {
+      MDSKey key = getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programId).build();
+      return getHistoricalRuns(key, status, startTime, endTime, limit, null, filter);
+    }
 
+    Predicate<MDSKey> keyPredicate = new AppVersionPredicate(ApplicationId.DEFAULT_VERSION);
+    MDSKey key = getVersionLessProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programId).build();
+    Map<ProgramRunId, RunRecordMeta> oldRecords = getHistoricalRuns(
+      key, status, startTime, endTime, limit, keyPredicate, filter);
+
+    int remaining = limit - oldRecords.size();
+    if (remaining > 0) {
+      // We need to scan twice since the key is modified again in getHistoricalRuns since we want to use the
+      // endTime and startTime to reduce the scan range
+      key = getProgramKeyBuilder(TYPE_RUN_RECORD_COMPLETED, programId).build();
+      Map<ProgramRunId, RunRecordMeta> newRecords = getHistoricalRuns(
+        key, status, startTime, endTime, remaining, keyPredicate, filter);
+      oldRecords.putAll(newRecords);
```

Comment:
I am not sure, but I will retain the new -> old ordering by switching the scan orders.